### PR TITLE
feat: add support for sessions in Renku 2.0

### PR DESCRIPTION
**Pull request by leafty**
See https://github.com/SwissDataScienceCenter/renku/issues/3475.

* Adds API endpoints for session environments
* Adds API endpoints for session launchers

/deploy #notest renku=release-0.50.x

### CODE REVIEW

File: components/renku_data_services/session/db.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+                )
+            return environment.dump()
+
+    async def insert_environment(
+        self,
+        user: base_models.APIUser,
+        new_environment: apispec.EnvironmentPost,
+    ) -> models.SessionEnvironment:
+        """Insert a new session environment."""
+        if user.id is None or not user.is_admin:
```
Comment (mana-alisafaee) [inline components/renku_data_services/session/db.py:53]: Should we allow creating envs for a specific project without being an admin? If yes, then it's ok to leave it for a future PR or story. Just make sure not to lose it.

Comment (mana-alisafaee) [reply components/renku_data_services/session/db.py:53]: Same goes for updating and deleting envs.

**Review by mana-alisafaee — CHANGES_REQUESTED** [review]
Thanks! Please see my comments.

File: components/renku_data_services/session/db.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+                res = await session.scalars(
+                    select(schemas.SessionLauncherORM).where(schemas.SessionLauncherORM.id == launcher_id)
+                )
+                launcher = res.one_or_none()
+
+                if launcher is None:
+                    return
+
+                authorized = await self.project_authz.has_permission(
+                    user=user, project_id=launcher.project_id, scope=Scope.WRITE
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/session/db.py:292]: Huh? Why? This is checking that the user has the `write` permission on the project.

File: components/renku_data_services/session/models.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+@dataclass(frozen=True, eq=True, kw_only=True)
+class Member(BaseModel):
+    """Member model."""
+
+    id: str
+
+
+@dataclass(frozen=True, eq=True, kw_only=True)
+class SessionEnvironment(BaseModel):
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/session/models.py]: Done in 68363c3067e19772b7d849c75979632d96eac6fe.

File: components/renku_data_services/session/db.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mana-alisafaee) [reply components/renku_data_services/session/db.py:290]: We have a special permission for delete which we check when deleting a project. Here should be the same I believe.

[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/session/db.py:290]: NO! I am removing a session launcher, not the project itself.

File: components/renku_data_services/session/db.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+    async def insert_launcher(
+        self, user: base_models.APIUser, new_launcher: apispec.SessionLauncherPost
+    ) -> models.SessionLauncher:
+        """Insert a new session launcher."""
+        if not user.is_authenticated or user.id is None:
+            raise errors.Unauthorized(message="You do not have the required permissions for this operation.")
+
+        project_id = new_launcher.project_id
+        authorized = await self.project_authz.has_permission(user=user, project_id=project_id, scope=Scope.READ)
```
Comment (leafty) [inline components/renku_data_services/session/db.py:173]:
```suggestion
        authorized = await self.project_authz.has_permission(user=user, project_id=project_id, scope=Scope.WRITE)
```

**Review by mana-alisafaee — APPROVED** [review]

**Pull request merged**